src/pipdeptree/_discovery.py

```diff
@@ -1,26 +1,44 @@
 from __future__ import annotations
 
+import ast
 import site
+import subprocess  # noqa: S404
 import sys
 from importlib.metadata import Distribution, distributions
+from pathlib import Path
 from typing import Iterable, Tuple
 
 from packaging.utils import canonicalize_name
 
 
 def get_installed_distributions(
+    interpreter: str = str(sys.executable),
     local_only: bool = False,  # noqa: FBT001, FBT002
     user_only: bool = False,  # noqa: FBT001, FBT002
 ) -> list[Distribution]:
     # See https://docs.python.org/3/library/venv.html#how-venvs-work for more details.
     in_venv = sys.prefix != sys.base_prefix
     original_dists: Iterable[Distribution] = []
+    py_path = Path(interpreter).absolute()
+    using_custom_interpreter = py_path != Path(sys.executable).absolute()
 
-    if local_only and in_venv:
+    if user_only:
+        original_dists = distributions(path=[site.getusersitepackages()])
+    elif using_custom_interpreter:
+        # We query the interpreter directly to get its `sys.path` list to be used by `distributions()`.
+        # If --python and --local-only are given, we ensure that we are only using paths associated to the interpreter's
+        # environment.
+        if local_only:
+            cmd = "import sys, site; print(site.getsitepackages([sys.prefix]))"
```

Comment:
Now that I think about it, it looks like that their is a potential to miss out on metadata when dealing with a custom interpreter or a virtual environment if we just use `site.getsitepackages()` . I think it makes sense to filter `sys.path` for environment-specific packages as most of the time `sys.path` already holds these site-packages directories.

We can filter `sys.path` by doing something like the following:
```suggestion
            cmd = "import sys; print([p for p in sys.path if p.startswith(sys.prefix)])"
```

Though it would make this PR not atomic, we should also do this in the `elif local_only and in_venv` branch.

`pip list` also does this by default when users pass `--local` and they're in a virtual env; see [here](https://github.com/pypa/pip/blob/31437d675abfa1563cffb3f68a0a23f3cbd68afa/src/pip/_internal/metadata/importlib/_envs.py#L170) (uses sys.path by default) and [here](https://github.com/pypa/pip/blob/31437d675abfa1563cffb3f68a0a23f3cbd68afa/src/pip/_internal/utils/misc.py#L368) (this function is used by `BaseDistribution.local`).

Reply:
It makes sense, done.

Reply:
Tests also need to adjust since we get path to search directly from `sys.path`.
Done this.